pkg/apis/logging/v1/output_types.go

```diff
@@ -0,0 +1,109 @@
+	// 'tls' are used for output types that don't define their own URL scheme.
+	// Example:
+	//
+	//     { type: syslog, destination: tls://syslog.example.com:1234 }
+	//
+	// TLS with server authentication is enabled by the URL scheme, for
+	// example 'tls' or 'https'.  See `secretRef` for TLS client authentication.
+	//
+	// +required
+	Destination string `json:"destination"`
+
+	// SecretRef refers to a `secret` object for secure communication.
+	//
+	// Client-authenticated TLS is enabled if the secret contains keys
+	// `tls.crt`, `tls.key` and `ca.crt`. Output types with password
+	// authentication will use keys `password` and `username`, not
+	// the exposed 'username@password' part of the `destination` URL.
+	//
+	// +optional
+	SecretRef corev1.SecretReference `json:"secret,omitempty"`
+
+	// Insecure must be true for intentionally insecure outputs.
+	// Has no function other than a marker to help avoid configuration mistakes.
+	//
+	// +optional
+	Insecure bool `json:"insecure,omitempty"`
+
+	// Reconnect configures how the output handles connection failures.
+	// Auto-reconnect is enabled by default.
+	//
+	// +optional
+	Reconnect *Reconnect `json:"reconnect,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=Unreliable;Retry
+type Reliability string
+
+const (
+	// Unreliable may drop data after a reconnect (at-most-once).
+	Unreliable = "Unreliable"
+
+	// Resend "in doubt" data after a reconnect. May cause duplicates (at-least-once).
+	// May enable buffering, blocking and/or acknowledgment features of the output type.
+	Resend = "Resend"
+)
+
+// Reconnect configures reconnect behavior after a disconnect.
+type Reconnect struct {
+	// FirstDelayMilliseconds is the time to wait after a disconnect before
+	// the first reconnect attempt. If reconnect fails, the delay is doubled
+	// on each subsequent attempt. The default is determined by the output type.
+	//
+	// +optional
+	FirstDelayMilliseconds int64 `json:"firstDelayMilliseconds,omitempty"`
+
+	// MaxDelaySeconds is the maximum delay between failed re-connect
+	// attempts, and also the maximum time to wait for an unresponsive
+	// connection attempt. The default is determined by the output type.
+	//
+	// +optional
+	MaxDelayMilliseconds int64 `json:"maxDelayMilliseconds,omitempty"`
+
+	// Reliability policy for data delivery after a re-connect.  This is
+	// simple short-hand for configuring the output to a given level of
+	// reliability.  The exact meaning depends on the output `type`.  The
+	// default is determined by the output type.
+	//
+	// +optional
+	Reliability Reliability `json:"reliability,omitempty"`
+}
+
+// OutputType is a union of supported output types.
+//
+// It always includes the required a `type` string and optionally a struct field
+// with type-specific configuration.
+type OutputType struct {
+	// Type of output plugin, for example 'syslog'
+	//
+	// +unionDiscriminator
+	// +required
+	Type string `json:"type"`
```

Comment:
Why do we need a string as union discriminator here? The types `Syslog` or whatever provide enough specificity, aren't they?. Following the k8s api for [corev1.Volume/VolumeSource](https://github.com/kubernetes/api/blob/master/core/v1/types.go#L38) here should be a good example.

Reply:
One argument would be that a given output is only one type. Which one is correct if multiple pointers are defined?

Reply:
@jcantrill I believe that's what the `+unionDiscriminator` is for.

I can see both sides of this -- we initially followed this pattern [1] but following the VolumeSource pattern makes this cleaner


[1] https://github.com/kubernetes/api/blob/master/apps/v1/types.go#L78

Reply:
@periklis there is a bunch of discussion on unions in the k8s world and contradictory examples, I *think* this is the "winning" pattern.  For our purposes I like it because it handles the empty case (which is our common case) better, i.e. I think this:
```
output:
  type: fluent
  url: tcp://foobar.org
```
Reads better than this:
```
output: 
  fluent: {}
  url: tcp://foobar.org
```